CoffeaAnalysis/HNLAnalysis/HNLAnalysis_tee.py

```diff
@@ -0,0 +1,334 @@
+            
+        if ds in self.stitched_list['WJets_samples']:
+            stitching_weights_WJets = import_stitching_weights('WJetsToLNu')
+            events = reweight_WJets(events, stitching_weights_WJets)
+        
+        # defining the mode
+        mode ='MCbackground' # default mode
+
+        if 'EGamma_2018' in ds:
+            # only keep the good runs
+            goodrun_lumi = data_goodrun_lumi(ds)
+            goodrun = goodrun_lumi[:,0]
+            events = events[np.isin(events.run, goodrun)]
+            events['genWeight'] = events.run > 0
+            ds = ds[0:-1]
+            mode ='Data'
+        
+        if 'HNL' in ds:
+            mode ='signal'
+        
+        out['sumw_all'][ds] += ak.sum(events.genWeight)
+        out['n_ev_all'][ds] += len(events)
+
+        # pileup correction: compute normalizing factor in order to keep the same number of events before and after correction (before any cut)
+        if mode != 'Data':
+            corr = get_pileup_correction(events, 'nominal')
+            self.norm_factor = ak.sum(events.genWeight)/ak.sum(events.genWeight*corr)
+
+        # MET filter folowing https://twiki.cern.ch/twiki/bin/viewauth/CMS/MissingETOptionalFiltersRun2#2018_2017_data_and_MC_UL
+        events = events[events.Flag.goodVertices & 
+                        events.Flag.globalSuperTightHalo2016Filter & 
+                        events.Flag.HBHENoiseFilter & 
+                        events.Flag.HBHENoiseIsoFilter & 
+                        events.Flag.EcalDeadCellTriggerPrimitiveFilter & 
+                        events.Flag.BadPFMuonFilter & 
+                        events.Flag.BadPFMuonDzFilter & 
+                        #events.Flag.hfNoisyHitsFilter & 
+                        events.Flag.eeBadScFilter & 
+                        events.Flag.ecalBadCalibFilter]
+
+        # Reco event selection: common minimal requirement for leptons
+        #tau
+        cut_tau_pt = 20. # Tau_pt > cut_tau_pt
+        cut_tau_eta = 2.5 #abs(Tau_eta) < cut_tau_eta
+        cut_tau_dz = 0.2 #abs(Tau_dz) < cut_tau_dz
+        cut_tau_idVSmu = 4 # idDeepTau2018v2p5VSmu >= Tight
+        cut_tau_idVSe = 3 # idDeepTau2018v2p5VSe >= VLoose
+        cut_tau_idVSjet = 5 # idDeepTau2018v2p5VSjet >= Medium
+    
+        #electrons
+        cut_e_pt = 25. # Electron_pt > cut_e_pt
+        cut_e_eta = 2.5 # abs(Electron_eta) < cut_e_eta
+        cut_e_dz = 0.2 #abs(Electron_dz) < cut_e_dz
+        cut_e_dxy = 0.045 # abs(Electron_dxy) < cut_e_dxy
+        cut_e_id = 0 # Electron_mvaIso_WP90 > cut_e_id (i.e True)
+        
+        #muons
+        cut_mu_pt = 25. # Muon_pt > cut_mu_pt
+        cut_mu_eta = 2.4 # abs(Muon_eta) < cut_mu_eta
+        cut_mu_dz = 0.2 #abs(Muon_dz) < cut_mu_dz
+        cut_mu_dxy = 0.045 # abs(Muon_dxy) < cut_mu_dxy
+        cut_mu_id = 0 #(Muon_mediumId > cut_mu_id || Muon_tightId > cut_mu_id)
+        cut_mu_iso = 0.15 # Muon_pfRelIso03_all < cut_mu_iso
+        
```

Comment:
here and for the other channels: define common selections outside if statements. E.g.
```python
tau_baseline = (events.Tau.pt > cut_tau_pt) & (np.abs(events.Tau.eta) < cut_tau_eta) & (np.abs(events.Tau.dz) < cut_tau_dz) & (events.Tau.idDeepTau2018v2p5VSmu >= cut_tau_idVSmu) & (events.Tau.idDeepTau2018v2p5VSe >= cut_tau_idVSe) & (events.Tau.decayMode != 5) & (events.Tau.decayMode != 6)
...
if self.region == 'A':
  events['SelTau'] = events.Tau[tau_baseline & (events.Tau.idDeepTau2018v2p5VSjet < cut_tau_idVSjet_high) & (events.Tau.idDeepTau2018v2p5VSjet >= cut_tau_idVSjet_low)]
  ...
```